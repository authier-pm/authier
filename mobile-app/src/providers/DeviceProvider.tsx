import React, { createContext } from 'react'
import { useForceUpdate } from '../../useForceUpdate'
import { device, Device } from '../utils/Device'

export const DeviceContext = createContext<Device>({} as any)

export const DeviceProvider = ({ children }) => {
  const forceUpdate = useForceUpdate()

  if (!device.fireToken) {
    device.initializePromise.then(() => {
      forceUpdate()
    })
    return null
  }

  return (
    <DeviceContext.Provider value={device}>{children}</DeviceContext.Provider>
  )
}
