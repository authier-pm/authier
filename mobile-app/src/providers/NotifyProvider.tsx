import AsyncStorage from '@react-native-async-storage/async-storage'
import React, {
  createContext,
  useState,
  useEffect,
  Dispatch,
  SetStateAction
} from 'react'

export const NotifyContext = createContext<{
  notifies: Array<Notifi>
  setNotifies: Dispatch<SetStateAction<Notifi[]>>
}>({} as any)

export interface Notifi {
  pageName: string
  device: string
  icon: string
  location: string
  time: string
}

export default function NotifyProvider({ children }) {
  const [notifies, setNotifies] = useState<Notifi[]>([])

  useEffect(() => {
    const getData = async () => {
      const jsonValue = await AsyncStorage.getItem('notifies', (e) => {
        if (e) console.log(e)
      })

      if (jsonValue) {
        let data = JSON.parse(jsonValue as string)
        setNotifies([data.data])
      } else {
        setNotifies([])
      }
    }

    getData()
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [])

  return (
    <NotifyContext.Provider value={{ notifies, setNotifies }}>
      {children}
    </NotifyContext.Provider>
  )
}
