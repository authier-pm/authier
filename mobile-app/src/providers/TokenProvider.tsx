import React, { createContext, useState, useEffect } from 'react'
import messaging from '@react-native-firebase/messaging'

export const GetTokenProvider = createContext<{ token: string | null }>(
  {} as any
)

export default function TokenProvider({ children }) {
  const [token, setToken] = useState<string | null>(null)

  useEffect(() => {
    async function getToken() {
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
      const refresed = await messaging().onTokenRefresh(async (fcm) => {
        console.log('test')
        setToken(fcm)
        return
      })
      const Token = await messaging().getToken()
      console.log('t', Token)
      setToken(Token)
    }

    getToken()
  }, [])

  return (
    <GetTokenProvider.Provider value={{ token }}>
      {children}
    </GetTokenProvider.Provider>
  )
}
