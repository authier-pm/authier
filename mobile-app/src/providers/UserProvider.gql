query isLoggedIn {
  authenticated
}

mutation logout {
  logout
}
