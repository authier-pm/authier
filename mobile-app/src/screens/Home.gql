query sendConfirmation($userId: String!, $success: Boolean!) {
  sendConfirmation(userId: $userId, success: $success)
}
