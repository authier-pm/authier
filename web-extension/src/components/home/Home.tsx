import React, {
  createContext,
  Dispatch,
  FunctionComponent,
  SetStateAction,
  useEffect,
  useState
} from 'react'

import { browser } from 'webextension-polyfill-ts'

import {
  Box,
  ChakraProvider,
  CircularProgress,
  Flex,
  Grid,
  Heading,
  useInterval,
  Text
} from '@chakra-ui/react'

import { Trans } from '@lingui/macro'

import { AddAuthSecretButton } from '../../popup/AddAuthSecretButton'
import { AuthsList } from '../../popup/AuthsList'
import { authenticator } from 'otplib'

export const Home: FunctionComponent = () => {
  const [seconds, setRemainingSeconds] = useState(authenticator.timeRemaining())

  useInterval(() => {
    setRemainingSeconds(authenticator.timeRemaining())
  }, 1000)

  return (
    <>
      <Flex position="sticky" align="center" p={4}>
        <CircularProgress
          min={1}
          max={30}
          value={30 - seconds}
          valueText={seconds.toString()}
          size="40px"
        />
        <AddAuthSecretButton />
        <Heading size="sm">
          <Trans>Logout</Trans>
        </Heading>
      </Flex>
      <Box height={200} width={300} p={5} mb={5}>
        <Grid gap={3} mb={5}>
          <AuthsList />
        </Grid>
      </Box>
    </>
  )
}
