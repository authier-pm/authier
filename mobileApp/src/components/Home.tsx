import {
  Avatar,
  Box,
  Button,
  FlatList,
  Flex,
  Icon as NativeIcon,
  IconButton,
  Modal,
  Text,
  Pressable,
} from 'native-base';
import React, { useEffect, useState } from 'react';
import { useContext } from 'react';
import ReactNativeBiometrics from 'react-native-biometrics';
import Icon from 'react-native-vector-icons/Ionicons';
import AsyncStorage from '@react-native-async-storage/async-storage';
import { NotifyContext } from '../NotifyProvider';

const Home = () => {
  const { notifies, setNotifies } = useContext(NotifyContext);
  const [isBio, setIsBio] = useState(false);
  const [open, setOpen] = useState(false);
  const [refreshing, setRefreshing] = useState(false);

  const authenticate = (page: string) => {
    if (isBio) {
      ReactNativeBiometrics.simplePrompt({
        promptMessage: 'Confirm fingerprint',
      })
        .then(async (resultObject) => {
          const { success } = resultObject;
          if (success) {
            //call here api, maybe wait on success confirmation

            setNotifies(
              notifies.filter((el) => (el.pageName === page ? false : true))
            );
            await AsyncStorage.removeItem('notifies', (e) => {
              if (e) console.log(e);
            });
          } else {
            console.log('user cancelled biometric prompt');
          }
        })
        .catch(() => {
          console.log('biometrics failed');
        });
    }
  };

  const updateList = async () => {
    const jsonValue = await AsyncStorage.getItem('notifies', (e) => {
      if (e) console.log('error in async storage', e);
    });

    if (jsonValue) {
      let data = JSON.parse(jsonValue as string);
      setNotifies([data.data]);
    } else {
      setNotifies([]);
    }

    setRefreshing(false);
  };

  useEffect(() => {
    async function getBio() {
      const { biometryType } = await ReactNativeBiometrics.isSensorAvailable();

      if (biometryType === ReactNativeBiometrics.Biometrics) {
        setIsBio(true);
      }
    }

    getBio();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  const handleRefresh = () => {
    setRefreshing(true);
    updateList();
  };

  //@ts-expect-error
  const ListItem = ({ item }) => {
    console.log('item', item);
    return (
      <Pressable
        key={item.pageName}
        backgroundColor="#ffffff"
        p={7}
        flexDirection="row"
        borderBottomWidth={0.5}
        borderBottomRadius={25}
        borderBottomColor="#a7a7a7"
        onPress={() => authenticate(item.pageName)}
      >
        <Flex>
          <Avatar
            size="lg"
            source={{
              uri: 'https://via.placeholder.com/150',
            }}
            mr={10}
          >
            NB
          </Avatar>
        </Flex>
        <Flex flexDirection="column">
          <Text fontSize={20}>{item.pageName}</Text>
          <Flex flexDirection="row" alignItems="center">
            <Icon name="desktop" color="gray" />
            <Text ml={1}>{item.device}</Text>
          </Flex>
          <Flex flexDirection="row" alignItems="center">
            <Icon name="locate" color="gray" />
            <Text ml={1}>{item.location}</Text>
          </Flex>
          <Flex flexDirection="row" alignItems="center">
            <Icon name="time" color="gray" />
            <Text ml={1}>{item.time}</Text>
          </Flex>
        </Flex>
        <Modal isOpen={open} onClose={() => setOpen(false)} mt={12}>
          <Modal.Content maxWidth={130} maxHeight={112}>
            <Modal.CloseButton />
            <Modal.Body>
              <Button.Group
                variant="ghost"
                display="flex"
                flexDirection="column"
              >
                <Button>Edit</Button>
                <Button>Delete</Button>
              </Button.Group>
            </Modal.Body>
          </Modal.Content>
        </Modal>
      </Pressable>
    );
  };

  return (
    <Box safeArea flex={1}>
      <Text>notification here</Text>

      <FlatList
        data={notifies}
        keyExtractor={(noti) => {
          return noti.pageName;
        }}
        renderItem={ListItem}
        onRefresh={handleRefresh}
        refreshing={refreshing}
      />
    </Box>
  );
};

export default Home;
