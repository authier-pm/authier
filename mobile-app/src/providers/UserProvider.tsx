import React, {
  createContext,
  useState,
  Dispatch,
  SetStateAction,
  useEffect
} from 'react'
import messaging from '@react-native-firebase/messaging'
import { useIsLoggedInQuery } from './UserProvider.codegen'
import { accessToken, clearAccessToken } from '../utils/tokenFromAsyncStorage'
import jwtDecode from 'jwt-decode'

export const UserContext = createContext<{
  isLogged: boolean
  setIsLogged: Dispatch<SetStateAction<boolean>>
  token: string | null
  isApiLoggedIn: Boolean
  logout: () => void
}>({} as any)

export default function UserProvider({ children }) {
  const [isLogged, setIsLogged] = useState(false)
  const [token, setToken] = useState<string | null>(null)
  const { data, loading } = useIsLoggedInQuery()

  useEffect(() => {
    async function getToken() {
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
      messaging().onTokenRefresh(async (fcm) => {
        console.log('ressetting token')
        setToken(fcm)
        return
      })
      const Token = await messaging().getToken()
      setToken(Token)
    }

    //Check asyncStorage if is the accessToken valid
    async function isAccessTokenValid() {
      const token = accessToken

      if (!token) {
        setIsLogged(false)
        return false
      }

      try {
        // @ts-expect-error
        const { exp } = jwtDecode(token)
        if (Date.now() >= exp * 1000) {
          setIsLogged(false)
          return false
        } else {
          setIsLogged(true)
          return true
        }
      } catch (error) {
        return false
      }
    }

    isAccessTokenValid()
    getToken()
  }, [data, loading])

  return (
    <UserContext.Provider
      value={{
        isLogged,
        setIsLogged,
        token,
        isApiLoggedIn: !!(data?.authenticated && !loading),
        logout: async () => {
          setIsLogged(false)
          clearAccessToken()
          // await logout()
        }
      }}
    >
      {children}
    </UserContext.Provider>
  )
}
