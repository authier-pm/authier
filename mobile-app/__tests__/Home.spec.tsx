import React from 'react'
import Home from '../src/screens/Home'
import { wrapInFEProviders } from '../test/providers'
import renderer from 'react-test-renderer'

it('Home component renders', () => {
  //const ac = makeSsrClient({})

  const tree = renderer.create(wrapInFEProviders(<Home />)).toJSON()
  expect(tree).toMatchSnapshot()
})
