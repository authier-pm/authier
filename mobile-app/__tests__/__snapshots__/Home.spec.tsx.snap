// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home component renders 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    Array [
      Object {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    flex={1}
    style={
      Array [
        Object {
          "flexBasis": 0,
          "flexGrow": 1,
          "flexShrink": 1,
        },
      ]
    }
  >
    <Text
      color="muted.700"
      fontSize="md"
      fontStyle="normal"
      fontWeight={400}
      style={
        Array [
          Object {
            "color": "#404040",
            "fontSize": 16,
            "fontStyle": "normal",
            "fontWeight": "400",
          },
        ]
      }
    >
      notification here
    </Text>
    <RCTScrollView
      disableVirtualization={false}
      getItem={[Function]}
      getItemCount={[Function]}
      horizontal={false}
      initialNumToRender={10}
      keyExtractor={[Function]}
      maxToRenderPerBatch={10}
      onContentSizeChange={[Function]}
      onEndReachedThreshold={2}
      onLayout={[Function]}
      onMomentumScrollEnd={[Function]}
      onRefresh={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      refreshControl={
        <RefreshControlMock
          onRefresh={[Function]}
          refreshing={false}
        />
      }
      refreshing={false}
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={50}
      stickyHeaderIndices={Array []}
      style={
        Array [
          Object {},
        ]
      }
      updateCellsBatchingPeriod={50}
      viewabilityConfigCallbackPairs={Array []}
      windowSize={21}
    >
      <RCTRefreshControl />
      <View />
    </RCTScrollView>
  </View>
</RNCSafeAreaProvider>
`;
